pkg/controller/machine.go

```diff
@@ -374,99 +334,180 @@ func (c *Controller) syncHandler(key string) error {
-				ownerRef := metav1.GetControllerOf(node)
-				if ownerRef == nil {
-					gv := machinev1alpha1.SchemeGroupVersion
-					node.OwnerReferences = []metav1.OwnerReference{*metav1.NewControllerRef(machine, gv.WithKind(machineKind))}
-					node, err = c.kubeClient.CoreV1().Nodes().Update(node)
-					if err != nil {
-						return fmt.Errorf("failed to update node %s after adding the owner ref: %v", node.Name, err)
-					}
-					glog.V(4).Infof("Added owner ref to node %s (machine=%s)", node.Name, machine.Name)
-					c.metrics.NodeJoinDuration.Observe(node.CreationTimestamp.Sub(machine.CreationTimestamp.Time).Seconds())
-				}
 
-				if node.Spec.ConfigSource == nil && machine.Spec.ConfigSource != nil {
-					node.Spec.ConfigSource = machine.Spec.ConfigSource
-					node, err = c.kubeClient.CoreV1().Nodes().Update(node)
-					if err != nil {
-						return fmt.Errorf("failed to update node %s after setting the config source: %v", node.Name, err)
+			glog.Infof("creating instance...")
+			if providerInstance, err = c.createProviderInstance(prov, machine, userdata); err != nil {
+				if ok, _, _ := cloudprovidererrors.IsTerminalError(err); ok {
+					message := fmt.Sprintf("%v. Unable to create a machine.", err)
+					if _, errNested := c.updateMachineError(machine, machinev1alpha1.CreateMachineError, message); errNested != nil {
+						return fmt.Errorf("failed to update machine error after failed machine creation: %v", errNested)
 					}
-					glog.V(4).Infof("Added config source to node %s (machine %s)", node.Name, machine.Name)
-				}
-				err = c.updateMachineStatus(machine, node)
-				if err != nil {
-					return fmt.Errorf("failed to update machine status: %v", err)
+					return err
 				}
+				return fmt.Errorf("failed to create machine at cloudprovider: %v", err)
+			}
+			// remove error message in case it was set
+			if machine, err = c.clearMachineErrorIfSet(machine); err != nil {
+				return fmt.Errorf("failed to update machine after removing the create machine error: %v", err)
 			}
+			glog.V(4).Infof("Created machine %s at cloud provider", machine.Name)
+			return nil
+		}
+
+		// case 1.2: terminal error was returned and manual interaction is required to recover
+		if ok, _, message := cloudprovidererrors.IsTerminalError(err); ok {
+			message = fmt.Sprintf("%v. Unable to create a machine.", err)
+			if _, errNested := c.updateMachineError(machine, machinev1alpha1.CreateMachineError, message); errNested != nil {
+				return fmt.Errorf("failed to update machine error after failed validation: %v", errNested)
+			}
+			return err
 		}
+
+		// case 1.3: transient error was returned requeue the request and try again in the future
+		return fmt.Errorf("failed to get instance from provider: %v", err)
 	}
 
-	if node != nil {
-		return c.ensureNodeLabelsAnnotationsAndTaints(node, machine)
+	// case 2: retrieving instance from cloudprovider was successfull
+	node, exists, err := c.getNode(providerInstance, string(providerConfig.CloudProvider))
```

Comment:
i would move the code which handles the kubernetes node object into a separate function